### ISSUE TITLE
Checking cellular modem behaviour/functionality

### ISSUE THREAD

**Issue by Bucknalla** [balena-io/device-diagnostics#153]
@xginn8 I'm interesting in contributing some tests to support validating cellular modem behaviour. Is that something considered in the realms of what device-diagnostics should support? 

If so, how would that behaviour be handled; would we want to enable cellular modem health checks via a flag or should it be something that the diagnostic script could extract from network/modem manager and ignore if it cannot find a cellular modem?

**Comment by xginn8** [comment]
@Bucknalla that would be wonderful! What sort of validation do you have in mind? The way this repo is structured there are both "checks" and "diagnostics". The former is used for expressing something to the user directly (some degraded state), while the latter is used to collect extra data for later investigation or for an expert to review.

It's a good question regardless. I just merged this week a check that is ignored if no hardware is available (https://github.com/balena-io/device-diagnostics/blob/master/scripts/checks.sh#L59-L75), so perhaps that's a jumping off point?

Let me know how I can help you!

**Comment by Bucknalla** [comment]
@xginn8 it would be great to report things like network strength, protocol (i.e. LTE/3G/etc.) maybe report the APN settings? I've often had devices that say they have network access but in reality what NetworkManager appears to show is just that the device has just registered on a network rather than actually able to TX/RX; a simple check for this would be great too.

My concern about ignoring the cellular modem is, how does this differentiate between there not being physical hardware for a cellular modem and if the cellular modem has completely failed?

**Comment by xginn8** [comment]
@Bucknalla I finally got around to playing with a modem locally. What I'd like to propose would be a check that does a few things:

If a modem is attached, we'll run a (dedicated) check that:
1. Confirms the modem is on
1. Returns the protocols available + configured APN
1. Asserts that the signal quality is above some threshold

If the OS detects no modem, that check will be turned off (which is gross, but a signal in and of itself).

We will also pull the descriptive output of `mmcli` into our diagnostics logs for further debugging.

Is there anything else you think is missing? The clunkiest bits for me is that this _probably_ belongs as a test in the larger `check_networking`, but since that masks successful tests, we'll want it to live outside of that check. Also the negation logic is weird, but I can't think of a better way.

**Comment by Bucknalla** [comment]
@xginn8 I agree, it feels like it should live in `check_networking` but I think this way does make sense. I'm wondering if there's a clean way for us to ping a server over the modem and to confirm that we have network access? Additionally I'm not sure what the logic is like for detecting the presence of SIM cards but that may also be interesting for the health check?
